### PR TITLE
Sim tau id es

### PR DESCRIPTION
**Pull request by conformist89**
Changes related to combined tau ID + es fit

### CODE REVIEW

File: combine_scripts/log_like_scan_poi.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+CHANNEL=$4
+WP=$5
+WP_ELE=$6
+
+
+
+categories=("DM0" "DM1" "DM10_11")
+
+
+tag_folder="medium_vs_ele_es_4_0"
```
Comment (harrypuuter) [inline combine_scripts/log_like_scan_poi.sh:16]: make script generic

File: combine_scripts/log_like_scan_poi.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+
+fix_es=0
+fix_id=0
+
+datacard_output_dm="datacards_es_4_0_29Apr_morph_v1/${NTUPLETAG}-${TAG}/${ERA}_tauid_${WP}"
```
Comment (harrypuuter) [inline combine_scripts/log_like_scan_poi.sh:22]: make scrip generic

File: combine_scripts/log_like_scan_poi.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    if [[ ! -d lscan_/${tag_folder}/${ERA}/${CHANNEL}/${TAG}/${dm_cat}/ ]]
+    then
+        mkdir -p lscan/${tag_folder}/${ERA}/${CHANNEL}/${WP}/${dm_cat}/
+        mkdir -p lscan/${tag_folder}/${ERA}/${CHANNEL}/${WP}/${dm_cat}/
+    fi
+
+    
+    if [[ $dm_cat == "DM0" ]]; then
+        fix_es=-0.999
```
Comment (harrypuuter) [inline combine_scripts/log_like_scan_poi.sh:35]: what are these values ?

File: combine_scripts/log_like_scan_poi.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    --cminFallbackAlgo Minuit2,Migrad,0:0.001 --cminFallbackAlgo Minuit2,Migrad,0:0.01 --cminPreScan \
+    --redefineSignalPOIs ES_${dm_cat},r \
+    --floatOtherPOIs=1 --points=400 --algo grid -v 2 -m 127
+    
+    
+    python3  plot1DScan.py  higgsCombine.${fit_name}.MultiDimFit.mH127.root --POI r \
+    --y-max 12 --output ${TAG}_${ERA}_${CHANNEL}_${WP}_${cat}_r_${dm_cat}
+    python3  plot1DScan.py  higgsCombine.${fit_name}.MultiDimFit.mH127.root --POI ES_${dm_cat} \
+     --y-max 12 --output ${TAG}_${ERA}_${CHANNEL}_${WP}_${cat}_es_${dm_cat}
+    mv ${TAG}_${ERA}_${CHANNEL}_${WP}_${cat}*  /work/olavoryk/tau_pog_tau_sfs/tau_id_es_main/smhtt_ul/lscan/${tag_folder}/${ERA}/${CHANNEL}/${WP}/${dm_cat}/
```
Comment (harrypuuter) [inline combine_scripts/log_like_scan_poi.sh:65]: make generic

File: config/shapes/channel_selection.py [hunk 5]
```diff
@@ -1,14 +1,38 @@
...
         ("extramuon_veto<0.5", "extramuon_veto"),
         ("dimuon_veto<0.5", "dilepton_veto"),
         ("q_1*q_2<0", "os"),
     ]
+    cuts = [
+        ("extraelec_veto<0.5", "extraelec_veto"),
+        ("extramuon_veto<0.5", "extramuon_veto"),
+        ("dimuon_veto<0.5", "dilepton_veto"),
+        ("q_1*q_2<0", "os"),
+    ]
```
Comment (harrypuuter) [inline config/shapes/channel_selection.py:17]: dublicated code

File: config/shapes/channel_selection.py [hunk 6]
```diff
@@ -222,10 +246,10 @@ def channel_selection(channel, era, special=None):
             cuts.extend(
                 [
                     ("id_tau_vsMu_Tight_2>0.5", "againstMuonDiscriminator"),
-                    ("id_tau_vsEle_VLoose_2>0.5", "againstElectronDiscriminator"),
-                    ("id_tau_vsJet_Medium_2>0.5", "tau_iso"),
+                    ("id_tau_vsEle_"+vs_ele_discr+"_2>0.5", "againstElectronDiscriminator"),
+                    ("id_tau_vsJet_"+vs_jet_discr+"_2>0.5", "tau_iso"),
```
Comment (harrypuuter) [inline config/shapes/channel_selection.py:250]: use f-strings instead of addition of strings

File: config/shapes/channel_selection.py [hunk 7]
```diff
@@ -325,3 +356,89 @@ def channel_selection(channel, era, special=None):
         return Selection(name="ee", cuts=cuts)
     else:
         raise ValueError("Given special selection does not exist")
+    
+
+def tau_id_channel_selection(channel, era, vs_jet_wp="tight", vs_ele_wp="vvloose"):
```
Comment (harrypuuter) [inline config/shapes/channel_selection.py:361]: whats the difference here compared to the `channel_selection` function ?

File: config/shapes/process_selection.py [hunk 8]
```diff
@@ -44,7 +153,25 @@ def prefiring_weight(era):
...
+    wps_dict = {
+
+        "vvtight" : "VVTight",
+        "vtight" : "VVTight",
+        "tight" : "Tight",
+        "medium" : "Medium",
+        "loose" : "Loose",
+        "vloose" : "VLoose",
+        "vvloose" : "VVLoose",
+        "vvvloose" : "VVVLoose",
```
Comment (harrypuuter) [inline config/shapes/process_selection.py:166]: why is this needed? Or crown naming are the ones with capital letters ? Why not use the same spelling for the shapes ?

File: config/shapes/process_selection.py [hunk 9]
```diff
@@ -346,16 +474,16 @@ def HWW_base_process_selection(channel, era):
...
 
 
-# def DY_process_selection(channel, era):
-#     DY_process_weights = DY_base_process_selection(channel, era).weights
+# def DY_process_selection(channel, era, vs_jet_wp, vs_ele_wp):
+#     DY_process_weights = DY_base_process_selection(channel, era, vs_jet_wp, vs_ele_wp).weights
 #     DY_process_weights.append((
 #         "((genbosonmass >= 50.0)*6.2139e-05*((npartons == 0 || npartons >= 5)*1.0 + (npartons == 1)*0.1743 + (npartons == 2)*0.3556 + (npartons == 3)*0.2273 + (npartons == 4)*0.2104) + (genbosonmass < 50.0)*numberGeneratedEventsWeight*crossSectionPerEventWeight)","z_stitching_weight"))
 #     return Selection(name = "DY",
 #                      weights = DY_process_weights)
```
Comment (harrypuuter) [inline config/shapes/process_selection.py:482]: why commented out?

File: config/shapes/process_selection.py [hunk 10]
```diff
@@ -20,6 +20,115 @@
...
+            "m_aiso_correction",
+        )
+    else:
+        weight = ("1.0", "m_aiso_correction")
+    return weight
```
Comment (harrypuuter) [inline config/shapes/process_selection.py:131]: This is all super old legacy code ??

File: config/shapes/variations.py [hunk 11]
```diff
@@ -790,14 +796,24 @@
...
+#     ReplaceWeight(
+#         "CMS_prefiringUp", "prefireWeight", Weight("prefiringweightup", "prefireWeight")
+#     ),
+#     ReplaceWeight(
+#         "CMS_prefiringDown",
+#         "prefireWeight",
+#         Weight("prefiringweightdown", "prefireWeight"),
+#     ),
+# ]
 prefiring = [
```
Comment (harrypuuter) [inline config/shapes/variations.py:809]: remove commented code

File: control_plots_ul.sh [hunk 12]
```diff
@@ -49,7 +49,7 @@ if [[ $MODE == "SHAPES" ]]; then
         --optimization-level 1 --control-plots \
         --control-plot-set ${VARIABLES} --skip-systematic-variations \
         --output-file $shapes_output \
-        --xrootd --validation-tag $TAG
+        --xrootd --validation-tag $TAG --vs-jet-wp medium --vs-ele-wp vvloose --special-analysis "TauID"
```
Comment (harrypuuter) [inline control_plots_ul.sh:52]: control plots should not use `--special-analysis "TauID`

File: emb_id_es_sep_cat.sh
Comment (harrypuuter) [inline emb_id_es_sep_cat.sh:1]: please put all these steering script into a seperate directory called e.g. `tauid_measurement`

File: emb_id_es_sep_cat.sh [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+
+# Datacard Setup
+
+datacard_output="datacards_test_pt_v3/${NTUPLETAG}-${TAG}/${ERA}_tauid_${WP}"
+
+datacard_output_dm="datacards_es_4_0_29Apr_morph_v1/${NTUPLETAG}-${TAG}/${ERA}_tauid_${WP}"
+
+datacard_output_incl="datacards_incl_test_v3/${NTUPLETAG}-${TAG}/${ERA}_tauid_${WP}"
+
+impact_path="impacts_test_v4"
```
Comment (harrypuuter) [inline emb_id_es_sep_cat.sh:26]: make generic

File: emb_id_es_sep_cat.sh [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+ "embminus2p3" "embminus2p2" "embminus2p1" "embminus2p0" "embminus1p9" "embminus1p8" "embminus1p7" "embminus1p6" "embminus1p5"\
+  "embminus1p4" "embminus1p3" "embminus1p2" "embminus1p1" "embminus1p0" "embminus0p9" "embminus0p8" "embminus0p7" "embminus0p6"\
+   "embminus0p5" "embminus0p4" "embminus0p3" "embminus0p2" "embminus0p1" "emb0p0" "emb0p1" "emb0p2" "emb0p3" "emb0p4" "emb0p5" "emb0p6"\
+    "emb0p7" "emb0p8" "emb0p9" "emb1p0" "emb1p1" "emb1p2" "emb1p3" "emb1p4" "emb1p5" "emb1p6" "emb1p7" "emb1p8" "emb1p9" "emb2p0" "emb2p1"\
+     "emb2p2" "emb2p3" "emb2p4" "emb2p5" "emb2p6" "emb2p7" "emb2p8" "emb2p9" "emb3p0" "emb3p1" "emb3p2" "emb3p3" "emb3p4" "emb3p5" "emb3p6"\
+      "emb3p7" "emb3p8" "emb3p9" "emb4p0")
+VS_ELE_WP="vvloose"
+
+
+dm_categories=("DM0" "DM1" "DM10_11")
```
Comment (harrypuuter) [inline emb_id_es_sep_cat.sh:70]: dublicated from above

File: emb_id_es_sep_cat.sh [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+min_es_sep=0
+max_es_sep=0
+
+cent_id_sep=0
+cent_es_sep=0
+
+id_fit_stri=""
+map_str=''
+
+mH=126
```
Comment (harrypuuter) [inline emb_id_es_sep_cat.sh:144]: make generic

File: emb_id_es_sep_cat.sh [hunk 16]
```diff
@@ -0,0 +1,443 @@
...
+    fit_categories=("DM0" "DM1" "DM10_11")
+
+    imp_es_dm0=-1
+
+    if [ ! -d "$impact_path/${ERA}/${CHANNEL}/${WP}" ]; then
+        mkdir -p  $impact_path/${ERA}/${CHANNEL}/${WP}
+    fi
+
+    
+    combineTool.py -M Impacts  -d ${WORKSPACE_IMP} -m 123 \
```
Comment (harrypuuter) [inline emb_id_es_sep_cat.sh:324]: why is it now `-m 123` and before its a variable ?

File: emb_tau_id_sfs_ul.sh
Comment (harrypuuter) [inline emb_tau_id_sfs_ul.sh:1]: same comments as above, make the script generic and put it into a seperate folder

File: emb_tau_id_sfs_ul.sh [hunk 17]
```diff
@@ -0,0 +1,666 @@
...
+if [[ $MODE == "MERGE" ]]; then
+    source utils/setup_root.sh
+    echo "[INFO] Merging outputs located in ${CONDOR_OUTPUT}"
+    hadd -j 5 -n 600 -f $shapes_rootfile ${CONDOR_OUTPUT}/../analysis_unit_graphs-${ERA}-${CHANNEL}-${NTUPLETAG}-${TAG}/*.root
+fi
+
+if [[ "${ERA}" == "2018"  ||  "${ERA}" == "2017" ]]; then 
+    datacard_era=${ERA}
+elif [[ "${ERA}" == "2016postVFP"  ||  "${ERA}" == "2016preVFP" ]]; then
+   datacard_era="2016"
```
Comment (harrypuuter) [inline emb_tau_id_sfs_ul.sh:188]: we do not destinguish between 2016 pre and post ?

File: emb_tau_id_sfs_ul.sh [hunk 18]
```diff
@@ -0,0 +1,666 @@
...
+        echo "[INFO] Input file: " output/$datacard_output_dm/htt_mt_${dm_cat}/higgsCombine.nominal_${dm_cat}.MultiDimFit.mH120.root
+        python3 plot_2D_scan.py --name nominal_${dm_cat} --in-path output/$datacard_output_dm/htt_mt_${dm_cat}/ \
+         --tau-id-poi ${dm_cat} --tau-es-poi ES_${dm_cat} --outname ${dm_cat} 
+    done
+
+fi
+
+
+fix_es=-1.4
+fix_id=1.097
```
Comment (harrypuuter) [inline emb_tau_id_sfs_ul.sh:356]: make generic

File: emb_tau_id_sfs_ul.sh [hunk 19]
```diff
@@ -0,0 +1,666 @@
...
+id_dm1=1.12
+es_dm1=-1.4
+
+
+min_id_dm10_11=0.9
+max_id_dm10_11=1.2
+min_es_dm10_11=-3
+max_es_dm10_11=1.5
+id_dm10_11=1.02
+es_dm10_11=-1
```
Comment (harrypuuter) [inline emb_tau_id_sfs_ul.sh:412]: see above (also this maybe needs some documentation what these parameters are used for)

File: friends/create_xpog_json_multifit.py [hunk 20]
```diff
@@ -0,0 +1,485 @@
...
+)
+
+
+correction_dm.generate_dm_scheme()
+
+# print("\n", correction_dm.correction)
+
+correctionset.add_correction(correction_pt)
+correctionset.add_correction(correction_dm)
+correctionset.write_json("Tau_"+str(args.wp)+"_"+str(args.era)+"UL_"+str(args.channel)+"_"+str(args.user_out_tag)+"_v1.json")
```
Comment (harrypuuter) [inline friends/create_xpog_json_multifit.py:485]: f-strings

File: friends/create_xpog_json_multifit.py
Comment (harrypuuter) [inline friends/create_xpog_json_multifit.py:1]: whats different in this file, compared to the existing version?

File: plot1DScan.py
Comment (harrypuuter) [inline plot1DScan.py:1]: whats different compared to https://github.com/KIT-CMS/smhtt_ul/blob/master/fitting/plot1DScan.py ?

Please put it in a folder instead of the main directory

**Review by harrypuuter — CHANGES_REQUESTED** [review]
Okay, quite some comments here, to quickly summarize the order to adress all this:

1. seperate out the produce_shapes.py for TauID into a seperate script and adapt all other script accoring to that
2. move all TauID related scripts into a seperate folder rather than the main directory, which is already too full imo.
3. remove unneeded / commented out code
4. see detailed comments

File: shapes/produce_shapes.py
[the remark this answers is not in the document]
Comment (conformist89) [reply shapes/produce_shapes.py:1]: I agree with that

File: config/shapes/channel_selection.py (hunk 7 above)
Comment (conformist89) [reply config/shapes/channel_selection.py:361]: I removed this code duplication

File: config/shapes/process_selection.py (hunk 9 above)
Comment (conformist89) [reply config/shapes/process_selection.py:482]: It is the leftover form `main` branch, there it's also commented, probably it's made for stitching but now we use inclusive samples only

File: config/shapes/process_selection.py (hunk 10 above)
Comment (conformist89) [reply config/shapes/process_selection.py:131]: Yes, we don't use them anymore, should I delete them?

File: emb_tau_id_sfs_ul.sh (hunk 17 above)
Comment (conformist89) [reply emb_tau_id_sfs_ul.sh:188]: We do and effectively we  get different scale factors for different eras, however morphing script doesn't differentiate between eras

File: friends/create_xpog_json_multifit.py
Comment (conformist89) [reply friends/create_xpog_json_multifit.py:1]: My version retrieves values from multidimensional fit

File: shapes/utils.py [hunk 21]
```diff
@@ -215,10 +215,15 @@ def add_tauES_datasets(
...
-                            )
+                            # cut.expression = cut.expression.replace(
+                            #     quant,
+                            #     "{quant}__{var}".format(quant=quant, var=shiftname),
+                            # )
+                            cut.expression = re.sub(
+                                rf"\b({quant})\b",
+                                f"{quant}__{shiftname}",
+                                cut.expression,
+                        )
```
[the remark this answers is not in the document]
Comment (conformist89) [reply shapes/utils.py:226]: To be able to proceed tau energy scales

File: config/shapes/channel_selection.py [hunk 22]
```diff
@@ -1,38 +1,33 @@
...
+
+    if vs_ele_wp not in wps_dict:
         print("This vs electron working point doen't exist. Please specify the correct vsEle discriminator ")
-    if vs_jet_wp not in wps_dict.keys():
-        print("This vs jet working point doen't exist. Please specify the correct vsEle discriminator ")
-    vs_ele_discr = wps_dict[vs_ele_wp]
-    vs_jet_discr = wps_dict[vs_jet_wp]
+    if vs_jet_wp not in wps_dict:
+        print("This vs jet working point doen't exist. Please specify the correct vsJet discriminator ")
+    vs_ele_discr = vs_ele_wp
```
Comment (harrypuuter) [inline config/shapes/channel_selection.py:29]: this redefinition is not needed, since you validate the WP anyhow, just use `vs_ele_wp` bwlog

File: config/shapes/channel_selection.py [hunk 23]
```diff
@@ -246,8 +241,8 @@ def channel_selection(channel, era, special=None, vs_jet_wp="tight", vs_ele_wp="
             cuts.extend(
                 [
                     ("id_tau_vsMu_Tight_2>0.5", "againstMuonDiscriminator"),
-                    ("id_tau_vsEle_"+vs_ele_discr+"_2>0.5", "againstElectronDiscriminator"),
-                    ("id_tau_vsJet_"+vs_jet_discr+"_2>0.5", "tau_iso"),
+                    ("id_tau_vsEle_%s_2>0.5" % (vs_ele_discr), "againstElectronDiscriminator"),
```
Comment (harrypuuter) [inline config/shapes/channel_selection.py:244]: use f-strings: `f"id_tau_vsEle_{vs_ele_discr}_2>0.5"`

Comment (harrypuuter) [reply config/shapes/channel_selection.py:244]: for all stuff below as well

File: control_plots_ul.sh [hunk 24]
```diff
@@ -42,14 +42,14 @@ if [[ $MODE == "SHAPES" ]]; then
...
-    python shapes/produce_shapes.py --channels $CHANNEL \
+    python shapes/produce_shapes_tauid_es.py --channels $CHANNEL \
         --directory $NTUPLES \
         --${CHANNEL}-friend-directory $XSEC_FRIENDS \
         --era $ERA --num-processes 4 --num-threads 12 \
         --optimization-level 1 --control-plots \
         --control-plot-set ${VARIABLES} --skip-systematic-variations \
         --output-file $shapes_output \
-        --xrootd --validation-tag $TAG --vs-jet-wp medium --vs-ele-wp vvloose --special-analysis "TauID" 
+        --xrootd --validation-tag $TAG --vs-jet-wp Medium --vs-ele-wp VVLoose --special-analysis "TauID"
```
Comment (harrypuuter) [inline control_plots_ul.sh:52]: control plot script should not to `--special-analysis "TauID` but regular control plots without any specific WPs and special analyses

File: plotting/plot_shapes_control.py [hunk 25]
```diff
@@ -78,7 +78,7 @@ def parse_arguments():
     parser.add_argument(
         "--tag",
         type=str,
-        default=None,
+        default=str,
```
Comment (harrypuuter) [inline plotting/plot_shapes_control.py:81]: this will crash - use and actual string (even if it is an empty one) `""`

File: plotting/plot_shapes_tauID_postfit.py [hunk 26]
```diff
@@ -363,7 +363,7 @@ def main(args):
                 except BaseException:
                     pass
             if channel != "mm":
-                plot.legend(i).add_entry(0, legend_bkg_processes[0], "#tau  "+legend_bkg_processes[0], "f")
+                plot.legend(i).add_entry(0, legend_bkg_processes[0], "#tau  %s" % (legend_bkg_processes[0]), "f")
```
Comment (harrypuuter) [inline plotting/plot_shapes_tauID_postfit.py:366]: f-strings

File: submit/submit_shape_production_ul.sh [hunk 27]
```diff
@@ -56,10 +57,10 @@ elif [[ "$SUBMIT_MODE" == "singlegraph" ]]; then
     echo "[INFO] Using friends $FRIENDS $NNSCORE_FRIENDS"
     [[ ! -d $OUTPUT ]] && mkdir -p $OUTPUT
     if [[ "$SPECIAL" == "TauID" ]]; then
-        python shapes/produce_shapes.py --channels $CHANNEL \
+        python shapes/produce_shapes_tauid_es.py --channels $CHANNEL \
```
Comment (harrypuuter) [inline submit/submit_shape_production_ul.sh:60]: the submit file should not do tauID stuff by default

**Review by harrypuuter — CHANGES_REQUESTED** [review]
some smaller remarks

**Comment by conformist89** [comment]
I think I implemented all the comments above

**Review by harrypuuter — APPROVED** [review]
